src/int_id.h

```diff
@@ -118,6 +120,54 @@ class int_id
         int _id;
 };
 
+// Fast set implementation for sequential int IDs. Only meant to be used with implementations based
+// on generic_factory, i.e. non-negative IDs that are sequential from 0.
+template <typename T, std::size_t kFastSize>
+class int_id_set
+{
+    public:
+        template <typename... Ts>
+        void emplace( Ts &&... ts ) {
+            insert( int_id<T>( std::forward<Ts>( ts )... ) );
+        }
+
+        void insert( int_id<T> id ) {
+            const std::size_t i = static_cast<std::size_t>( id.to_i() );
+            if( i < fast_set_.size() ) {
+                fast_set_[i] = true;
+            } else {
+                slow_set_.insert( id );
+            }
+        }
+
+        void erase( int_id<T> id ) {
+            const std::size_t i = static_cast<std::size_t>( id.to_i() );
+            if( i < fast_set_.size() ) {
+                fast_set_[i] = false;
+            } else {
+                slow_set_.erase( id );
+            }
+        }
+
+        bool contains( int_id<T> id ) const {
+            const std::size_t i = static_cast<std::size_t>( id.to_i() );
+            if( i < fast_set_.size() ) {
+                return fast_set_[i];
+            } else {
+                return slow_set_.count( id );
+            }
+        }
+
+        void clear() {
+            fast_set_.reset();
+            slow_set_.clear();
+        }
+
+    private:
+        std::bitset<kFastSize> fast_set_;
```

Comment:
Didn't know we rolled our own, but I think we should stick to this one for now. It lets the fast path size be encoded in the type so all uses can be updated at once, and contains is called a lot. It makes up 5% of the profile alone, just for monster flags (that's not including the str->int conversion, which was even more on top), so I'd prefer as little indirection and branching as possible. That said I'm obviously not super attached to this, so we can just replace it in the future. I still have a lot of stuff coming down the pipe.

I notice that you aren't using my favorite bitset trick though... so now I need to find a problem I can shoe horn it into.

Reply:
I can probably easily make the inline storage size configurable for tiny bit set.

Reply:
That's one option that could work, although I don't want to block the PR on the change. It's easy enough to swap later.

Reply:
Eh ok, I threw it together but even in the fast path, with tiny_bitset it adds a branch+mask/shift (to get the current size of the bitset) compared to int_id_set contains as-is, because the size is a compile constant there. Since it's just one per mtype the size savings aren't worth it.